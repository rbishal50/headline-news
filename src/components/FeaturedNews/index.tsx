import { FC, useEffect, useState } from 'react'
import { useSelector, useDispatch, shallowEqual } from 'react-redux'

import { RootState } from '../../redux/rootReducer'
import { fetchFeatured } from '../../redux/slices/newsSlice'
import ListItems from '../ListItems'
import PrimaryCard from '../PrimaryCard'

interface IProps {
  clicked: (data: any) => void
}

export interface INews {
  id: string
  title: string
  subtitle: string
  content: string
  imageUrl: string
}

export const HeadlineNews: FC<IProps> = ({ clicked }) => {
  const [highLightedNews, setHighLightedNews] = useState<INews>()
  const dispatch = useDispatch()

  const { featured } = useSelector((state: RootState) => {
    return {
      featured: state.news.featured,
    }
  }, shallowEqual)

  useEffect(() => {
    dispatch(fetchFeatured())
  }, [dispatch])

  useEffect(() => {
    if (featured.length) {
      setHighLightedNews(featured[0])
    }
  }, [featured])

  return (
    <>
      {highLightedNews && (
        <PrimaryCard
          title={highLightedNews.title}
          clicked={clicked}
          subtitle={highLightedNews.subtitle}
          imageUrl={highLightedNews.imageUrl}
          size='sm'
        />
      )}
      <ListItems items={featured} clicked={clicked} />
    </>
  )
}

export default HeadlineNews
