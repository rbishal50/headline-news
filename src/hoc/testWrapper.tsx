import { ComponentType, ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'

import store from '../redux/store'

const testWrapper = (Component: any, props = {}): ReactElement => {
  return (
    <BrowserRouter>
      <Provider store={store}>
        <Component {...props} />
      </Provider>
    </BrowserRouter>
  )
}

export default testWrapper
