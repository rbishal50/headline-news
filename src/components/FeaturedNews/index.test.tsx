import { render } from '@testing-library/react'
import * as reactRedux from 'react-redux'

import testWrapper from '../../hoc/testWrapper'
import FeaturedNews from './index'

const mockFunction = jest.fn()

describe('<FeaturedNews />', () => {
  const useSelectorMock = jest.spyOn(reactRedux, 'useSelector')

  beforeEach(() => {
    useSelectorMock.mockClear()
  })

  it('renders items', () => {
    useSelectorMock.mockReturnValue({
      featured: [
        { id: '1', title: 'test title 1' },
        { id: '2', title: 'test title 2', subtitle: 'subtitle 2' },
      ],
    })
    const { container } = render(testWrapper(FeaturedNews, { clicked: mockFunction }))
    expect(container).toMatchSnapshot()
  })
})
