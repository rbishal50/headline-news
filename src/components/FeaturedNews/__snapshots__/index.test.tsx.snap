// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeaturedNews /> renders items 1`] = `
<div>
  <div
    class="primary-card mb-4 flex-column card"
    data-testid="card"
  >
    <div>
      <img
        alt="Cover"
        class="card-img cover"
        src="https://picsum.photos/500/300"
      />
    </div>
    <div>
      <div
        class="card-body"
        style="max-width: 500px;"
      >
        <div
          class="card-title card-title h5"
          style="font-size: 20px;"
        >
          test title 1
        </div>
      </div>
    </div>
  </div>
  <div
    class="list-items list-group list-group-flush"
  >
    <div
      class="list-group-item"
    >
      <span
        class="title"
      >
        test title 1
      </span>
       
      <p
        class="text-muted mb-0 subtitle"
      />
    </div>
    <div
      class="list-group-item"
    >
      <span
        class="title"
      >
        test title 2
      </span>
       
      <p
        class="text-muted mb-0 subtitle"
      >
        subtitle 2
      </p>
    </div>
  </div>
</div>
`;
